File: java/src/main/java/com/android/widget/RecyclerView/AATest/TestMultiAdapter.java
package com.android.widget.RecyclerView.AATest;

import android.content.Context;
import com.android.java.R;
import com.android.util.ToastUtil;
import com.android.widget.RecyclerView.AATest.entity.DetailBean;
import com.android.widget.RecyclerView.AATest.entity.MonthBean;
import com.android.widget.RecyclerView.BaseAdapter;
import com.android.widget.RecyclerView.MultiViewType;
import com.android.widget.RecyclerView.ViewHolder;
import com.google.gson.Gson;

import java.util.List;

/**
 * Create by xuzhb on 2020/1/21
 * Desc:多布局
 */
public class TestMultiAdapter extends BaseAdapter<String> {

    public TestMultiAdapter(Context context, List<String> dataList) {
        super(context, dataList, new MultiViewType<String>() {
            @Override
            public int getLayoutId(String data, int position, int totalCount) {
                if (dataList.get(position).contains("month")) {
                    return R.layout.item_month;
                }
                return R.layout.item_detail;
            }
        });
    }

    @Override
    protected void bindData(ViewHolder holder, String data, int position) {
        Gson gson = new Gson();
        if (data.contains("month")) {
            MonthBean bean = gson.fromJson(data, MonthBean.class);
            holder.setText(R.id.month_tv, bean.getMonth()).
                    setText(R.id.count_tv, "总计" + bean.getCount() + "个");
        } else {
            DetailBean bean = gson.fromJson(data, DetailBean.class);
            holder.setText(R.id.detail_tv, bean.getDetail()).setText(R.id.time_tv, bean.getTime())
                    .setOnItemClickListener(v -> {
                        ToastUtil.showToast(bean.getDetail());
                    });
        }
    }
}

File: kotlin/src/main/java/com/android/widget/RecyclerView/AATest/TestMultiAdapter.kt
package com.android.widget.RecyclerView.AATest

import android.content.Context
import com.android.basicproject.R
import com.android.util.ToastUtil
import com.android.util.isEntity
import com.android.widget.RecyclerView.AATest.entity.DetailBean
import com.android.widget.RecyclerView.AATest.entity.MonthBean
import com.android.widget.RecyclerView.BaseAdapter
import com.android.widget.RecyclerView.MultiViewType
import com.android.widget.RecyclerView.ViewHolder
import com.google.gson.Gson

/**
 * Created by xuzhb on 2019/10/2
 * Desc:多布局
 */
class TestMultiAdapter(context: Context, list: MutableList<String>) :
    BaseAdapter<String>(context, list, object : MultiViewType<String> {
        override fun getLayoutId(item: String, position: Int, totalCount: Int): Int {
            return when {
//                isEntity(list[position], MonthBean::class) -> R.layout.item_month
                list[position].contains("month") -> R.layout.item_month
                else -> R.layout.item_detail
            }
        }
    }) {
    override fun bindData(holder: ViewHolder, data: String, position: Int) {
        when {
            data.contains("month") -> {
                val bean: MonthBean = Gson().fromJson(data, MonthBean::class.java)
                holder.setText(R.id.month_tv, bean.month).setText(R.id.count_tv, "总计${bean.count}个")
            }
            else -> {
                val bean: DetailBean = Gson().fromJson(data, DetailBean::class.java)
                holder.setText(R.id.detail_tv, bean.detail).setText(R.id.time_tv, bean.time)
                holder.setOnItemClickListener {
                    ToastUtil.showToast(bean.detail)
                }
            }
        }
    }

}